cou/apps/base.py

```diff
@@ -607,6 +609,82 @@ def _get_charmhub_migration_step(self, target: OpenStackRelease) -> PreUpgradeSt
             ),
         )
 
+    def get_run_deferred_hooks_and_restart_pre_upgrade_step(self) -> list[PreUpgradeStep]:
+        """Get the steps for run deferred hook and restart services for before upgrade.
+
+        :return: Steps for run deferred hooks and restart service
+        :rtype: List of PreUpgradeStep
+        """
+        units = self.get_units_to_run_action()
+        run_hook_step = PreUpgradeStep(
+            description=(
+                "Execute run-deferred-hooks for all units to clear any leftover events:"
+                f" {', '.join([unit.name for unit in units])}"
+            ),
+            parallel=False,
+        )
+        run_hook_step.add_steps(
```

Comment:
I don't see the advantage of adding heirarchy here?

Reply:
it's cleaner in the plan

```
                        Execute run-deferred-hooks for all units to clear any leftover events: rabbitmq-server/6, rabbitmq-server/7, rabbitmq-server/8
                                Execute run-deferred-hooks on unit: 'rabbitmq-server/6'
                                Execute run-deferred-hooks on unit: 'rabbitmq-server/7'
                                Execute run-deferred-hooks on unit: 'rabbitmq-server/8'
```

Reply:
Ah ok that's fair.  Could you remove the unit listing after `leftover events:` though?  That will make the line shorter.  The units are already listed underneath.

Reply:
Thanks, I was thinking the same

Reply:
Done